### ISSUE TITLE
Sort order from bash not matching scala > or compareto

### ISSUE THREAD

**Issue by kurzum** [dbpedia/databus-maven-plugin#44]
The code for comparing one line with the previous in scala is:
https://github.com/dbpedia/databus-maven-plugin/blob/master/src/main/scala/org/dbpedia/databus/lib/Datafile.scala#L217

@Vehnem @JJ-Author Could you check the details on how this works? The goal is that scala matches to whatever we are doing with sort in bash. At the moment this is not the case. I attached a file sorted with LC_ALL=C on download server, scala says it is not sorted. I ran the stuff below on the download server. On my machine both LC_ALL=C and LC_COLLATE=C work fine. 

Could you create a small test file with 5 or 6 entries, which boil down the problem?

```
root@vmdbpedia:/media/bigone/releases/databus-maven-plugin/dbpedia/mappings# bzcat geo-coordinates-mappingbased/src/main/databus/2018.09.12/geo-coordinates-mappingbased_id.ttl.bz2 | LC_COLLATE=C sort -c 
sort: -:13: disorder: <http://id.dbpedia.org/resource/Aa_en_Hunze> <http://www.georss.org/georss/point> "52.95 6.6" .
root@vmdbpedia:/media/bigone/releases/databus-maven-plugin/dbpedia/mappings# bzcat geo-coordinates-mappingbased/src/main/databus/2018.09.12/geo-coordinates-mappingbased_id.ttl.bz2 | LC_ALL=C sort -c 
```

**Comment by kurzum** [comment]
`uni2ascii -p -a A <file>`

**Comment by kurzum** [comment]
```
A_Fam
Aa_en 
```

```
<U0041><U005F><U0046><U0061><U006D>
<U0041><U0061><U005F><U0065><U006E>
```

**Comment by kurzum** [comment]
```
 echo -n "" > sorttest ; for i in {20..79}; do  echo "\u$i"| ascii2uni -q -a U >> sorttest ; done ;  for i in {100..599}; do  echo "\u$i"| ascii2uni -q -a U >> sorttest ; done ; for i in {607..1000}; do  echo "\u$i"| ascii2uni -q -a U >> sorttest ; done
```

**Comment by kurzum** [comment]
from dataid.ttl mappings/geo-coordinates
```
<#geo-coordinates-mappingbased-2018.09.12_id.ttl.bz2>
        a                            dataid:SingleFile ;
        rdfs:label                   "Geo Coordinates Mappingbased"@en ;
        dataid:associatedAgent       <https://webid.dbpedia.org/webid.ttl#this> ;
        dataid:compression           "bzip2" ;
        dataid:contentVariant        "id" ;
        dataid:duplicates            "0"^^xsd:decimal ;
        dataid:formatExtension       "ttl" ;
        dataid:isDistributionOf      <#geo-coordinates-mappingbased-2018.09.12> ;
        dataid:maintainer            <https://termilion.github.io/webid.ttl#this> ;
        dataid:nonEmptyLines         "22860"^^xsd:decimal ;
        dataid:preview               "<http://id.dbpedia.org/resource/1_Decembrie> <http://www.georss.org/georss/point> \"44.28333333333333 26.066666666666666\" .\n<http://id.dbpedia.org/resource/1_Decembrie> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2003/01/geo/wgs84_pos#SpatialThing> .\n<http://id.dbpedia.org/resource/1_Decembrie> <http://www.w3.org/2003/01/geo/wgs84_pos#lat> \"44.28333333333333\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<http://id.dbpedia.org/resource/1_Decembrie> <http://www.w3.org/2003/01/geo/wgs84_pos#long> \"26.066666666666666\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<http://id.dbpedia.org/resource/23_August,_Constanᘺ> <http://www.georss.org/georss/point> \"43.92 29.965\" .\n<http://id.dbpedia.org/resource/23_August,_Constanᘺ> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2003/01/geo/wgs84_pos#SpatialThing> .\n<http://id.dbpedia.org/resource/23_August,_Constanᘺ> <http://www.w3.org/2003/01/geo/wgs84_pos#lat> \"43.92\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<http://id.dbpedia.org/resource/23_August,_Constanᘺ> <http://www.w3.org/2003/01/geo/wgs84_pos#long> \"29.965\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<http://id.dbpedia.org/resource/A_Famosa> <http://www.georss.org/georss/point> \"2.1916166666666665 102.25030555555556\" .\n<http://id.dbpedia.org/resource/A_Famosa> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2003/01/geo/wgs84_pos#SpatialThing> ." ;
        dataid:sha256sum             "3e6f79130c1b02e13232696b911baabdacd3a6d5a416f4ed630ed4883ef66eb4" ;
        dataid:signature             "fV2qJp9Av5E1y/EiNW+lcnz4yPxKMpcrYo4ks++ZbFlxiTXQQMNxlRGXyDwa626LSD4IfBnAE1EbBEjYld8/8LiUVM4GcAS151knZvVH3xyjvN83dvXRQ39/kh95OkVdKYyintvAAb9tipwfl37EX7zqNL+eUaVNkHkrGLFY91LhkH0mhyxtTrWTWXq0vHJOjID+657FQJCdJcRev3/XbIrO8xnaaHdggvTPGIQR2qPXj4Z6DnZLCZ9clxqqYn01/4LWEjNg9ON25B+dubNl+g2GUQuWyTIGlF4Uqftmm6w00saBxRtJx69pPJyb3+gZF7/jM1Jhyr64sddWQNTvMA==" ;
        dataid:sorted                false ;
```

**Comment by kurzum** [comment]
checked in sorttest in examples/animals/fish:
```
<#fish-2018.08.15_sorttest.txt>
        a                            dataid:SingleFile ;
        rdfs:label                   "Infobox Properties"@en ;
        dataid:associatedAgent       <https://neradis.github.io/test-id.ttl#this> ;
        dataid:compression           "None" ;
        dataid:contentVariant        "sorttest" ;
        dataid:duplicates            "0"^^xsd:decimal ;
        dataid:formatExtension       "" ;
        dataid:isDistributionOf      <#fish-2018.08.15> ;
        dataid:nonEmptyLines         "953"^^xsd:decimal ;
        dataid:preview               " \n!\n\"\n#\n$\n%\n&\n'\n(\n)" ;
        dataid:sha256sum             "0c6aea0c88e73c1e8eac66ae21211ba0e2b9634ee9665344ca8b294bf1b82684" ;
        dataid:signature             "LxuLji3FKz2LnxE1tAYNGXJd0rGzOKvMwgSGDgAULd9ifdXXsMNCufZNPqiW1H1Rj3J9hk5vwEZczwYjgIBT+WL8LAmFrUZM13WtUkPx4e27Jmn//dN2NBXl6UkdWJZptZD4txbUzyaES90MK+hBuTSCwrEQmDjmgMaK33e+dSB3xdm8n2NjdDd3RrBQiipQl7jtQ+zSM1WqUGwzKcunpixQSqlo+q9zgcTLr9QI6JXR1GcytHnDKBi1AmLhLJ/lcdZJMB2Dpe1UmowQf7Ojw4OHthVcffUU0Iw26PvYSUofJHUrVkm1CuBzDWUZ+GPTXoamQGQa76OlrFcMf1gcjw==" ;
        dataid:sorted                true ;
```
so scala `compareto` recognizes the first 1000 chars correctly

**Comment by JJ-Author** [comment]
the example where we found out that utf8 sort causes issues in deduplication was the following:

```
<http://dbpedia.org/resource/Dacalana_liaoi> <http://dbpedia.org/ontology/thumbnail> <http://commons.wikimedia.org/wiki/Special:FilePath/Dacalana_liaoi_♂_upperside.jpg?width=300> .
<http://dbpedia.org/resource/Dacalana_liaoi> <http://dbpedia.org/ontology/thumbnail> <http://commons.wikimedia.org/wiki/Special:FilePath/Dacalana_liaoi_♀_upperside.jpg?width=300> .
<http://dbpedia.org/resource/Dacalana_liaoi> <http://xmlns.com/foaf/0.1/depiction> <http://commons.wikimedia.org/wiki/Special:FilePath/Dacalana_liaoi_♂_upperside.jpg> .
<http://dbpedia.org/resource/Dacalana_liaoi> <http://xmlns.com/foaf/0.1/depiction> <http://commons.wikimedia.org/wiki/Special:FilePath/Dacalana_liaoi_♀_upperside.jpg> .
```

**Comment by kurzum** [comment]
LC_ALL=C order:
```
♀
♂
```
```
<U2640>
<U2642>
```

**Comment by kurzum** [comment]
added to example/animals/fish/sorttest

**Comment by kurzum** [comment]
[log.txt](https://github.com/dbpedia/databus-maven-plugin/files/2749338/log.txt)

**Comment by JJ-Author** [comment]
scala code
```
val m = "♂"
val f = "♀"
println(m>f);println(m<f)
```

output
```
m: String = ♂
f: String = ♀
true
res0: Unit = ()
false
res1: Unit = ()
```

**Comment by JJ-Author** [comment]
```
root@vmdbpedia:/media/bigone/releases/databus-maven-plugin/dbpedia/mappings# bzcat geo-coordinates-mappingbased/src/main/databus/2018.09.12/geo-coordinates-mappingbased_id.ttl.bz2 | LC_COLLATE=C sort -c 
sort: -:13: disorder: <http://id.dbpedia.org/resource/Aa_en_Hunze> <http://www.georss.org/georss/point> "52.95 6.6" .

```

**Comment by JJ-Author** [comment]
so there are 2 conclusions: LC_ALL=C sort -u is safer since independend of custom locale settings
second the file seems correctly byte sorted.

**Comment by JJ-Author** [comment]
OK i added some more logging and run the scala parts of the tool:

```
object test {
  def main(args: Array[String]): Unit = {
    val df = new Datafile(new java.io.File("/home/kilt/Downloads/geo-coordinates-mappingbased_id.ttl.bz2"))(null)
    df.updateFileMetrics()
  }
}

```

the output is: 

```
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    Sort order wrong   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<http://id.dbpedia.org/resource/Sle> <http://www.w3.org/2003/01/geo/wgs84_pos#long> "12.933333333333334"^^<http://www.w3.org/2001/XMLSchema#float> .
<http://id.dbpedia.org/resource/S𐏎le> <http://www.georss.org/georss/point> "45.62 25.709722222222222" .

```

**Comment by JJ-Author** [comment]
tried to fix in 8005c6654a276e1b96b7216342f591ae31e84f50

**Comment by kurzum** [comment]
fixed, proof is here: http://downloads.dbpedia.org/temporary/log.txt (for a while )
